Profile API should show node details as well as shard details

quux00 opened the pull request:
SearchProfileResults has additional fields for XContent output: node_id, cluster, index, shard_id. It parses the existing composite ID using the new parseProfileShardId method, which reverses the SeachShardTarget.toString method.

No new information is added here, merely the splitting out of the four pieces of information in the profile shards "composite" id that is created by the SeachShardTarget.toString method.

Profile/shards output now has the form:
```
  "profile": {
    "shards": [
      {
        "id": "[2m7SW9oIRrirdrwirM1mwQ][blogs][0]",
        "node_id": "2m7SW9oIRrirdrwirM1mwQ",
        "shard_id": "0",
        "index": "blogs",
        "cluster": "(local)",
        "searches": [ ... ]
        ...
      },
      {
        "id": "[UngEVXTBQL-7w5j_tftGAQ][remote1:blogs][2]",
        "node_id": "UngEVXTBQL-7w5j_tftGAQ",
        "shard_id": "2",
        "index": "blogs",
        "cluster": "remote1",
        "searches": [ ... ]
        ...
```

where the latter is on a remote cluster and you can see that as the prefix on the index name.

Partially addresses #25896

quux00 committed:
Profile API should show node details as well as shard details

SearchProfileResults has additional fields for XContent output: node_id, cluster, index, shard_id.
It parses the existing composite ID using the new parseProfileShardId method, which reverses
the SeachShardTarget.toString method.

No new information is added here, merely the splitting out of the four pieces of information
in the profile shards "composite" id that is created by the SeachShardTarget.toString method.

Profile/shards output now has the form:
  "profile": {
    "shards": [
      {
        "id": "[2m7SW9oIRrirdrwirM1mwQ][blogs][0]",
        "node_id": "2m7SW9oIRrirdrwirM1mwQ",
        "shard_id": "0",
        "index": "blogs",
        "cluster": "(local)",
        "searches": [ ... ]
        ...
      },
      {
        "id": "[UngEVXTBQL-7w5j_tftGAQ][remote1:blogs][2]",
        "node_id": "UngEVXTBQL-7w5j_tftGAQ",
        "shard_id": "2",
        "index": "blogs",
        "cluster": "remote1",
        "searches": [ ... ]
        ...

where the latter is on a remote cluster and you can see that as the prefix on the index name.

Partially addresses #25896

quux00 committed:
Properly updated profile.asciidoc with all embedded test snippets passing

quux00 committed:
code comment cleanup

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticsearchmachine commented:
Hi @quux00, I've created a changelog YAML for you.

quux00 committed:
Fixed javadoc for record ShardProfileId

quux00 commented on server/src/main/java/org/elasticsearch/search/profile/SearchProfileResults.java:
```diff
@@ -79,9 +90,24 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
             builder.field(ID_FIELD, key);
-            shardResults.get(key).toXContent(builder, params);
+
+            ShardProfileId shardProfileId = parseProfileShardId(key);
```
Hi Ben,

> I think this is good for backwards compatibility, but what do you think of actually serializing those objects back instead of encoding them in the id and parsing them out again? It seems brittle to me to store important info in a plain text string that is assumed to be formatted in a specific way.

In terms of brittleness, I marked this PR as partially addressing the original issue: https://github.com/elastic/elasticsearch/issues/25896, where the request was basically to split the composite ID and maybe add some additional information. They recommended not changing the ID itself because that would be a backwards incompatible change.

So the brittleness is already baked in when that composite ID was created and now we need to support that composite ID in order to not make breaking changes.

In my [analysis comment on that issue](https://github.com/elastic/elasticsearch/issues/25896#issuecomment-1565087817), I analyzed all the formats of that ID and how it's generated, indicating that it would be fairly straightforward to parse that out. Adding new info is problematic though because that composite ID is used as a Map key (https://github.com/elastic/elasticsearch/pull/96396/files#diff-582e9063bae3a8d070700f30a160c2647fd69bff489591ce5a16821bac63af6aR52) and that Map is passed around quite a bit through multiple layers before it gets to the place where it is rendered as XContent. IIRC (it's been two weeks since I looked at this), the likely place to serialize the data is in [SearchShardTarget](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/search/SearchShardTarget.java), but that doesn't make out to the place the XContent is rendered (only the composite ID map key).

I think we'd need to refactor a fair bit of code to avoid using that ID as the map key. So I decided to try to go the simple path first - just parsing out the existing info from that map key.

We could certainly try the bigger refactoring if we think it's worth it and want additional things not in the composite id, like `cluster_uuid`.


> the brittleness could be acceptable if test coverage was adequate.

In terms of better test coverage, the approach I took was bottom up from the possible formats that are generated by [SearchShardTarget.toString()](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/search/SearchShardTarget.java#L117), which is the composite ID map key.

I can enhance [the test I added](https://github.com/elastic/elasticsearch/pull/96396/files#diff-14544e5bfd8cc0411e14ddfff362b9862f32eeb25879f94ada3d7ed6f8f8fadaR35) to use the SearchShardTarget.toString() method to generate the input.

I'm open to suggestions on how else to test it. Since the node_id and possibly other fields would be randomized in yamlRestTests. The test would need to parse out the composite on the fly and I'm not sure how I'd do that in a yamlRestTest. Do you have ideas on an additional good way to test this?

benwtrent commented on server/src/main/java/org/elasticsearch/search/profile/SearchProfileResults.java:
```diff
@@ -173,4 +199,53 @@ private static void parseProfileResultsEntry(XContentParser parser, Map<String,
+            }
+            return new ShardProfileId(nodeId, indexName, shardId, cluster);
+        } else {
+            logger.warn("Unable to match input against expected pattern of [nodeId][indexName][shardId]. Input: {}", compositeId);
```
Could you add an assertion here so this is caught if it fails during testing? We should never enter this `else` block.

quux00 replied:
Good idea. I need to get used to adding asserts!

benwtrent reviewed (commented):
Thanks for the clarifications @quux00 

Could you add some tests on the output and add an `assert` to the parser to catch during tests?

Here is where you should put the integration tests:
rest-api-spec/test/search/370_profile.yml

These yaml tests are ran against various Elasticsearch versions. So you will have to restrict the version via 

```
  - skip:
      version: " - 8.7.99"
      reason: introduced in 8.8.0
```
Or whatever version is appropriate.

quux00 committed:
PR feedback

Added asserts to the SearchProfileResults.parseCompositeProfileShardId method to assert against invalid inputs

Added yamlRestTest for the new fields in the profile response.

Updated the existing unit test to use more randomization of inputs and to use the
SearchShardTarget.toString method to generate the compositeId (which is what happens in production)

quux00 committed:
SearchProfileResultsTests.createTestItem now properly uses SearchShardTarget.toString to set the key for the SearchProfileShardResult map

elasticsearchmachine commented:
Hi @quux00, I've created a changelog YAML for you.

benwtrent commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/370_profile.yml:
```diff
@@ -401,3 +401,30 @@ dfs profile for search with dfs_query_then_fetch:
+---
+composite_id_parsed:
+  - skip:
+      version: " - 8.8.99"
```
With 8.9.0 being cut, you will need to update this as it will get released in 8.10 now.

quux00 replied:
Fixed. Thanks for the reminder.

benwtrent replied:
I suppose you haven't pushed your local commit yet? Once you do, I will give it another review

quux00 replied:
Pushed it now. It must have failed when I tried it earlier and I missed it.

benwtrent commented on server/src/main/java/org/elasticsearch/search/profile/SearchProfileResults.java:
```diff
@@ -170,4 +196,53 @@ private static void parseProfileResultsEntry(XContentParser parser, Map<String,
+    static ShardProfileId parseCompositeProfileShardId(String compositeId) {
+        assert Strings.isNullOrEmpty(compositeId) == false : "An empty id should not be passed to parseCompositeProfileShardId";
+
+        Pattern r = Pattern.compile("\\[([^]]+)\\]\\[([^]]+)\\]\\[(\\d+)\\]");
```
Can we make this a `private static final Pattern SHARD_ID_DECOMPOSITION = Pattern.compile("\\[([^]]+)\\]\\[([^]]+)\\]\\[(\\d+)\\]");`

For every profile result, we could have many different IDs and while profiling is already slow as it bypasses cache and adds additional overhead, lets not needlessly create short lived objects and recompile a regex pattern.

quux00 replied:
Fixed.

benwtrent reviewed (commented):
two minor things and I think this is good to go.

quux00 committed:
PR feedback - updated yaml test skip to 8.9.99 and changed compositeShardId regex to private static final var

benwtrent reviewed (approved):
:shipit:

quux00 committed:
when running when with remote cluster the cluster changes, so can't match on known value
